provider/oauth2/views.py

```diff
@@ -22,12 +20,10 @@ class OAuth2AccessTokenMixin(AccessTokenMixin):
     def get_access_token(self, request, user, scope, client):
         try:
             # Attempt to fetch an existing access token.
-            at = AccessToken.objects.get(user=user, client=client,
-                                         scope=scope, expires__gt=now())
+            at = AccessToken.objects.get(user=user, client=client, scope=scope, expires__gt=now())
         except AccessToken.DoesNotExist:
             # None found... make a new one!
             at = self.create_access_token(request, user, scope, client)
-            self.create_refresh_token(request, user, scope, at, client)
```

Comment:
This change makes sense since not all grant types should create a refresh token.
However, what about the grant types that DO require a refresh token?
Specifically, [authorization_code](https://github.com/edx/django-oauth2-provider/blob/edx/provider/views.py#L570) and the non-Public [password](https://github.com/edx/django-oauth2-provider/blob/edx/provider/views.py#L604) grants also call `get_access_token` - expecting to have created a refresh token.

In fact, looking at the code now, there's an unintentional bug with Public clients using the Password Grant with constants.SINGLE_ACCESS_TOKEN enabled - the code generated refresh tokens for them.

My suggestion to you is to use the recommendation in the [Wrong Abstraction](http://www.sandimetz.com/blog/2016/1/20/the-wrong-abstraction) article.  Have each of the 3 callers of the `get_access_token` method handle the `DoesNotExist` exception themselves.  So, `authorization_code` and `password` for private Clients will create refresh tokens, while `client_credentials` and password for public Clients won't.  (Alternatively, pass a boolean into `get_access_token` - ignoring the article.)